### PR TITLE
Space permissions

### PR DESCRIPTION
**Pull request by greatbacon**
This PR introduces another 5 permissions, these ones related to Spaces.

### CODE REVIEW

File: src/lib/app/templates.ts [hunk 1]
```diff
@@ -85,7 +85,14 @@ export const defaultStandardCommunityRoles: RoleTemplate[] = [
 	{
 		name: 'Member',
 		default: true,
-		policies: [{permission: 'InviteToCommunity'}],
+		//TODO how to do this without raw strings & instead use permissions?
```
Comment (ryanatkn) [inline src/lib/app/templates.ts:88]: add `} as const;` to line 16 of `permissions.ts`, then we should be able to do `permissions.InviteToCommunity` etc

File: src/lib/vocab/policy/permissions.ts [hunk 2]
```diff
@@ -8,5 +8,10 @@ export const permissions = {
 	DeleteRole: 'DeleteRole',
 	CreateAssignment: 'CreateAssignment',
 	DeleteAssignment: 'DeleteAssignment',
+	CreateSpace: 'CreateSpace',
+	ReadSpace: 'ReadSpace',
+	ReadSpaces: 'ReadSpaces',
+	UpdateSpace: 'UpdateSpace',
+	DeleteSpace: 'DeleteSpace',
 };
```
Comment (ryanatkn) [inline src/lib/vocab/policy/permissions.ts:16]:
```suggestion
} as const;
```

File: src/lib/app/templates.ts [hunk 3]
```diff
@@ -85,7 +85,14 @@ export const defaultStandardCommunityRoles: RoleTemplate[] = [
...
+		policies: [
+			{permission: 'InviteToCommunity'},
+			{permission: 'CreateSpace'},
+			{permission: 'ReadSpace'},
+			{permission: 'ReadSpaces'},
+			{permission: 'UpdateSpace'},
```
Comment (ryanatkn) [inline src/lib/app/templates.ts:94]:
```suggestion
			{permission: permissions.InviteToCommunity},
			{permission: permissions.CreateSpace},
			{permission: permissions.ReadSpace},
			{permission: permissions.ReadSpaces},
			{permission: permissions.UpdateSpace},
```

see the other comments -- also you'll need to add `permissions` to the imports

File: src/lib/vocab/space/spaceServices.ts [hunk 4]
```diff
@@ -116,8 +126,14 @@ export const UpdateSpaceService: ServiceByName['UpdateSpace'] = {
...
 		transact(async (repos) => {
 			const {space_id, actor: _actor, ...partial} = params;
-			const space = unwrap(await repos.space.update(space_id, partial));
-			return {ok: true, status: 200, value: {space}};
+			const space = unwrap(await repos.space.findById(space_id));
+			if (!space) {
+				return {ok: false, status: 404, message: 'no space found'};
+			}
+
+			unwrap(await checkPolicy(permissions.UpdateSpace, _actor, space.community_id, repos));
```
Comment (ryanatkn) [inline src/lib/vocab/space/spaceServices.ts:134]:
```suggestion
			unwrap(await checkPolicy(permissions.UpdateSpace, actor, space.community_id, repos));
```

File: src/lib/vocab/space/spaceServices.ts [hunk 5]
```diff
@@ -116,8 +126,14 @@ export const UpdateSpaceService: ServiceByName['UpdateSpace'] = {
 	perform: ({transact, params}) =>
 		transact(async (repos) => {
 			const {space_id, actor: _actor, ...partial} = params;
```
Comment (ryanatkn) [inline src/lib/vocab/space/spaceServices.ts:128]:
```suggestion
			const {space_id, actor, ...partial} = params;
```

File: src/lib/vocab/policy/permissions.ts [hunk 6]
```diff
@@ -8,5 +8,10 @@ export const permissions = {
 	DeleteRole: 'DeleteRole',
 	CreateAssignment: 'CreateAssignment',
 	DeleteAssignment: 'DeleteAssignment',
+	CreateSpace: 'CreateSpace',
+	ReadSpace: 'ReadSpace',
+	ReadSpaces: 'ReadSpaces',
```
Comment (ryanatkn) [inline src/lib/vocab/policy/permissions.ts:13]: I think `ReadSpaces` should be deleted

**Review by ryanatkn — APPROVED** [review]
looks great, one comment I think we should talk about, I don't have an answer yet: https://github.com/feltcoop/felt-server/pull/661#discussion_r1070415352

**Comment by greatbacon** [comment]
> looks great, one comment needs discussion, I don't have an answer yet: [#661 (comment)](https://github.com/feltcoop/felt-server/pull/661#discussion_r1070415352)

We have it on our docket to discuss stuff around spaces & permissions, so I'll hold off on resolving this question & merging the PR until then.

**Comment by greatbacon** [comment]
For future reference, we agreed to remove the singular form, & are adopting a implicit `checkCommunityAccess` permission for some of these read functions. Also we fixed some broken tests, yay!

**Pull request merged**